### PR TITLE
Metric-Learning

### PR DESCRIPTION
**Pull request by AlekseySh**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md), Pull Request section?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: catalyst/contrib/utils/misc.py [hunk 1]
```diff
@@ -56,4 +56,18 @@ def args_are_not_none(*args: Optional[Any]) -> bool:
     return True
 
 
-__all__ = ["args_are_not_none", "make_tuple", "pairwise"]
+def find_all(x: Iterable[Any], x0: Any) -> List[int]:
```
Comment (Scitator) [inline catalyst/contrib/utils/misc.py]: is it possible to find more easy-to-go name? this one looks too general

File: catalyst/data/sampler.py [hunk 2]
```diff
@@ -76,6 +81,99 @@ def __len__(self) -> int:
...
+        assert (1 < p <= len(classes)) and (1 < k)
+        assert all(
+            [n > 1 for n in Counter(labels).values()]
+        ), "Each class shoud contain at least 2 instances to fit (1)"
+
+        self._labels = tuple(labels)  # to make it immutable
+        self._p = p
+        self._k = k
+
+        self._bs = self._p * self._k
```
Comment (Scitator) [inline catalyst/data/sampler.py]: could you please use full naming, like `_batch_size`?

File: catalyst/data/sampler.py [hunk 3]
```diff
@@ -76,6 +81,99 @@ def __len__(self) -> int:
...
+        self._classes = classes
+
+        # to satisfy statement (1)
+        n_classes = len(self._classes)
```
Comment (Scitator) [inline catalyst/data/sampler.py]: `num_classes ` maybe? it would be more appropriate for Catalyst codestyle

File: catalyst/data/sampler.py [hunk 4]
```diff
@@ -76,6 +81,99 @@ def __len__(self) -> int:
...
+        if n_classes % self._p == 1:
+            self._n_epoch_classes = n_classes - 1
```
Comment (Scitator) [inline catalyst/data/sampler.py]: the same here, `_num_epoch_classes ` maybe?

File: catalyst/data/sampler.py (hunk 2 above)
Comment (AlekseySh) [reply catalyst/data/sampler.py]: done

File: catalyst/data/sampler.py (hunk 4 above)
Comment (AlekseySh) [reply catalyst/data/sampler.py]: done

File: catalyst/data/sampler.py (hunk 3 above)
Comment (AlekseySh) [reply catalyst/data/sampler.py]: done

File: catalyst/contrib/utils/misc.py (hunk 1 above)
Comment (AlekseySh) [reply catalyst/contrib/utils/misc.py]: done

File: catalyst/contrib/utils/misc.py [hunk 5]
```diff
@@ -56,4 +56,18 @@ def args_are_not_none(*args: Optional[Any]) -> bool:
     return True
 
 
-__all__ = ["args_are_not_none", "make_tuple", "pairwise"]
+def find_value_ids(x: Iterable[Any], x0: Any) -> List[int]:
```
Comment (Scitator) [inline catalyst/contrib/utils/misc.py]: maybe `def find_value_ids(it: Iterable[Any], value: Any) -> List[int]:`?

File: catalyst/data/tests/test_sampler.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    num_classes_in_data = len(set(labels))
+    num_classes_in_epoch = len(set(sampled_classes))
+    assert (num_classes_in_data == num_classes_in_epoch) or (
+        num_classes_in_data == num_classes_in_epoch + 1
+    )
+
+    assert max(sampled_ids) <= len(labels) - 1
+
+
+def test_balancec_batch_sampler(input_balancec_batch_sampler):
```
Comment (Scitator) [inline catalyst/data/tests/test_sampler.py]: `test_balance_batch_sampler`? looks like typo

File: catalyst/data/tests/test_sampler.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        p, k = randint(2, 12), randint(2, 12)
+        labels_ = [[label] * randint(2, 12) for label in range(p + 1)]
+        labels = [el for sublist in labels_ for el in sublist]
+        shuffle(labels)
+        input_cases.append((labels, p, k))
+
+    return input_cases
+
+
+def single_check_balancec_batch_sampler(
```
Comment (Scitator) [inline catalyst/data/tests/test_sampler.py]: `single_check_balance_batch_sampler`? typo?

File: catalyst/data/tests/test_sampler.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+from operator import itemgetter
+from random import randint, shuffle
+
+import pytest
+
+from catalyst.data.sampler import BalancecBatchSampler
+
+
+@pytest.fixture()
+def input_balancec_batch_sampler() -> List[Tuple[List[int], int, int]]:
```
Comment (Scitator) [inline catalyst/data/tests/test_sampler.py]: `input_balance_batch_sampler`?

**Comment by Scitator** [comment]
btw, could you please update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?

File: catalyst/data/tests/test_sampler.py (hunk 6 above)
Comment (AlekseySh) [reply catalyst/data/tests/test_sampler.py]: yes, thx

File: catalyst/data/tests/test_sampler.py (hunk 7 above)
Comment (AlekseySh) [reply catalyst/data/tests/test_sampler.py]: thx, fixed

File: catalyst/data/tests/test_sampler.py (hunk 8 above)
Comment (AlekseySh) [reply catalyst/data/tests/test_sampler.py]: thx, fixed

File: catalyst/contrib/utils/misc.py (hunk 5 above)
Comment (AlekseySh) [reply catalyst/contrib/utils/misc.py]: done

**Comment by AlekseySh** [comment]
> btw, could you please update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?

**Comment by AlekseySh** [comment]
miss click, I did'n want to close the PR)

**Comment by AlekseySh** [comment]
CHANGELOG - done

**Review by Scitator — APPROVED** [review]

**Comment by Scitator** [comment]
@AlekseySh could you please fix the codestyle?
`catalyst-make-codestyle && catalyst-check-codestyle`

**Pull request merged**